byterun/minor_gc.c

```diff
@@ -607,25 +613,21 @@ void caml_empty_minor_heap_domain (struct domain* domain)
     caml_ev_end("minor_gc/ephemerons");
 
     caml_ev_begin("minor_gc/update_minor_tables");
-    for (r = minor_tables->major_ref.base;
-         r < minor_tables->major_ref.ptr; r++) {
+    for (r = minor_tables->major_ref.base, r_new = minor_tables->major_ref_rewrites.base;
+         r < minor_tables->major_ref.ptr;
+         r++, r_new++) {
+      CAMLassert (r_new < minor_tables->major_ref_rewrites.ptr);
       value v = **r;
       if (Is_block (v) && is_in_interval ((value)Hp_val(v), young_ptr, young_end)) {
-        value vnew;
-        header_t hd = Hd_val(v);
-        int offset = 0;
-        if (Tag_hd(hd) == Infix_tag) {
-          offset = Infix_offset_hd(hd);
-          v -= offset;
-        }
-        CAMLassert (Hd_val(v) == 0);
-        vnew = Op_val(v)[0] + offset;
+        value vnew = (value)*r_new;
         CAMLassert (Is_block(vnew) && !Is_minor(vnew));
         CAMLassert (Hd_val(vnew));
```

Comment:
I think these assertions are wrong. I think it's possible for a Forward_tag block to contain an immediate rather than a block (e.g. `lazy (2 + 2)`). If the GC runs after it's forced, it will short-circuit references to the block into the immediate `Val_int(4)`, which will fail the `Is_block(vnew)` test.

A correct assertion here would be:

`!Is_block(vnew) || (!Is_minor(vnew) && Hd_val(vnew) != 0)`

Reply:
good spot. I've updated the testcase to exercise this path and also updated the assert.